src/services/download/mod.rs

```diff
@@ -34,26 +34,21 @@ pub use crate::types::ObjectId;
 /// HTTP User-Agent string to use.
 const USER_AGENT: &str = concat!("symbolicator/", env!("CARGO_PKG_VERSION"));
 
-#[derive(Debug, Fail, Clone)]
-pub enum DownloadErrorKind {
-    #[fail(display = "failed to download")]
-    Io,
-    #[fail(display = "bad file destination")]
-    BadDestination,
-    #[fail(display = "failed writing the downloaded file")]
-    Write,
-    #[fail(display = "download was cancelled")]
+#[derive(Debug, Error)]
+pub enum DownloadError {
+    // TODO: this is currently an opaque `failure::Error`, as we pack all kinds of errors into this
+    #[error("failed to download")]
+    Io(failure::Error),
```

Comment:
Would it work if you make this `Box<dyn Error + 'static>` or even `Box<dyn Error + Send + Sync + 'static>`? 

Also, consider implementing a utility function for `map_err`:

```rust
impl DownloadError {
    fn io<E>(error: E) -> Self 
    where
        E: Error + 'static
    {
        Self::Io(Box::new(error))
    }
}

some_action().map_err(DownloadError::io)?
```

Reply:
Alternatively we're piling too many things into one variant and should use more variants?

Reply:
I introduced a separate `GcsError` and split up the `DownloadError::Io` into better variants.

The problem is that quite some foreign error types, most notably those from actix_web don’t `impl std::error::Error` and other traits such as `Clone`, etc :-(